src/library_gl.js

```diff
@@ -251,8 +251,8 @@ var LibraryGL = {
     getSource: function(shader, count, string, length) {
       var source = '';
       for (var i = 0; i < count; ++i) {
-        var len = length ? {{{ makeGetValue('length', 'i*4', 'i32') }}} : undefined;
-        source += UTF8ToString({{{ makeGetValue('string', 'i*4', 'i32') }}}, len >= 0 ? len : undefined);
+        var len = length ? {{{ makeGetValue('length', 'i*4', 'i32') }}} : -1;
+        source += UTF8ToString({{{ makeGetValue('string', 'i*4', 'i32') }}}, len < 0 ? 9e9/*short abbrv. for infinity*/ : len);
```

Comment:
I think this would be clearer with `Math.infinity`, and we can assume closure will minify it (if not, we can file a bug, and could temporarily force it to do so with a toplevel `var Math_infinity = Math.infinity` and use `Math_infinity` here.

Reply:
It would be clearer with infinity, but similarly outlining `var Math_infinity = Infinity;` would be shorter only if multiple uses of `Math_infinity` could amortize the hit. (also all uses of such var `Math_infinity` should likely have to be in cold program paths, because a JIT can see that `Infinity` is always a constant literal, whereas a dereference via a `Math_infinity` could sneakily change under it, so it could not do JIT magic on such a variable)

9e9 is pretty large, no? :)

Reply:
It seems safer to add `Math_infinity` and use that - otherwise `9e9` will eventually get us in trouble, or at least be confusing to read - it just looks arbitrary.

Another option may be to do `var VERY_LARGE = 9e9;` with a big comment on that, and use that in places like this. Or is this really the only one?

Reply:
That code site has several `'i32` accesses right next to it, so I am not concerned e.g. about wasm64 or similar. This is currently the only place in code that has a construct like this.

`var VERY_LARGE = 9e9;` is a good idea, but we cannot do it until https://github.com/google/closure-compiler/issues/3187 is resolved. (Having a function local define and use of `VERY_LARGE` is practically about the same as `9e9/*VERY_LARGE*/`)

Reply:
If we can't rely on closure for such an optimization, another option is to write one ourselves. https://github.com/emscripten-core/emscripten/pull/7973 has a new estree-based optimization infrastructure, which is much nicer to write passes in (has a proper AST, instead of uglify1's `node[1][3][2]` etc.).

Reply:
We cannot write a generic optimizer pass to optimize `Infinity` to `9e9`, because we don't know about the use case, In this specific location `9e9` is large enough to serve as infinity, but in other places it may not be?

Reply:
Oh you meant specifically for the literal expansion of a `VERY_LARGE` variable.. I am not sure that it would be something I can jump on right now, and the worth is not that high, especially since it would be to automatically optimize what was trivial to write in code in the first place.